{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step definition"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step options\n",
    "\n",
    "**Step options** are specified after step name that assists the specification of workflows. SoS provides the following options"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `skip`\n",
    "\n",
    "Option `skip` takes two formats, the first format has no value\n",
    "\n",
    "```\n",
    "[10: skip]\n",
    "```\n",
    "\n",
    "and is equivalent to\n",
    "\n",
    "```\n",
    "[10: skip=True]\n",
    "```\n",
    "\n",
    "The whole step will be skipped as if it is not defined at all in the script. This option provides a quick method to disable a step.\n",
    "\n",
    "The second format takes a value, which is usually an expression that will be evaluated when the step is executed. For example, step 10 is by default executed by default"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "default_10 is executed\n"
     ]
    }
   ],
   "source": [
    "parameter: qc = True\n",
    "[10 (quality check): skip=not qc]\n",
    "print(\"${step_name} is executed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "but will not be executed if the workflow is executed with option `--no-qc` (for `qc=False`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%run --no-qc\n",
    "parameter: qc = True\n",
    "[10 (quality check): skip=not qc]\n",
    "print(\"${step_name} is executed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `sigil`\n",
    "\n",
    "Option `sigil` accepts a string for an alternative sigil, or `None` to disable string interpolation in the step. The sigil must be two strings separated by a space, such as `%( )`, `< >`, and `#{ }`. Sigils with equal left and right symbol such as `# #` can be used although they do not support features such as nested interpolation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `shared`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SoS executes each step in a separate process and by default does not return any result to the master SoS process. Option `shared` is used to share variables between steps. This option accepts:\n",
    "\n",
    "* A string (variable name), or\n",
    "* A map between variable names and expressions (strings) that will be evaluated upon the completion of the step.\n",
    "* A sequence of strings (variables) or maps.\n",
    "\n",
    "After the completion of a step, local variables defined in the `shared` option will be sent to the global environment so that later steps can use the updated value. For example,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Argument input_files of type <class 'list'> is required"
     ]
    }
   ],
   "source": [
    "parameter: trim = False\n",
    "parameter: input_files = list\n",
    "\n",
    "[10: shared={'input_files':'output'}, skip=trim is False]\n",
    "input:  input_files, group_by=1\n",
    "output: '${_input[0]}.trimmed'\n",
    "\n",
    "sh:\n",
    "    trim ${input} > ${output}\n",
    "\n",
    "[20]\n",
    "input: input_files\n",
    "# do something"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example, `trim` is an option to trim input files defined in another parameter `input_files`. If `trim` is false, step 10 is ignored and step 20 handles `input_files`. If `trim` is True, the files are trimmed and produces `${_input[0]}.trimmed`. Variable `input_files` are sent back to the global environment so that step 20 would be handling the trimmed version of input files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A map syntax is recommended to share `output` of one step with others, because the variable assignment will be evaluated only after the step is complete:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "[1: shared = {'test_output': 'output'}]\n",
    "...\n",
    "[2]\n",
    "input: test_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The map syntax is evaluated as expressions; therefore it is possible to finer control what specific output, or variations of output, to share with others. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "[1: shared={'test_output_1':'output[0]', 'test_output_2': 'output[1]'}]\n",
    "...\n",
    "[2]\n",
    "input: test_output_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "to shared the first file in `output` (filename `output[0]`) instead of the entire output file list.\n",
    "\n",
    "NOTE: Because global variables are readonly, they cannot be changed by individual steps through the `shared` option. Shared variable therefore has to be either new variables or parameters.\n",
    "\n",
    "The `shared` option also provides a `sos_variable` target. Please refer to section [auxiliary step](Auxiliary_steps.html) for details."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `provides`\n",
    "\n",
    "This option lists files or targets a step generates so that it can be called if the target is required but does not exist. Please see section [auxiliary step](Auxiliary_Steps.html) for details. A step can be treated as both a forward step and an auxiliary step despite of the differences on how they are invoked."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `workdir`\n",
    "\n",
    "Option `workdir` changes the workdir of the current step to specified directory. It also creates the directory if it does not already exist.\n",
    "\n",
    "For example, step 10 in the following example creates directory `tmp` and changes working directory to it before creating a file `a.txt`. The change of directory, however, does not affect how variables are interpreted so `output` is still `a.txt`, not `tmp/a.txt`, so `${myoutput}` does not point to the correct file in the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a.txt: False\n"
     ]
    }
   ],
   "source": [
    "%sandbox\n",
    "[10: workdir='tmp', shared={'myoutput': 'output[0]'}]\n",
    "output: 'a.txt'\n",
    "run:\n",
    "    touch ${output}\n",
    "\n",
    "[20]\n",
    "print(\"${myoutput}: ${os.path.isfile(myoutput)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Step input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step Input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input files\n",
    "\n",
    "The input of SoS step follows the following rules:\n",
    "\n",
    "* **the input of a SoS step is by default the output of the previous step**, which is `None` for the first step.\n",
    "* **`input` option**, which could be **a list** of filenames (string literal, variables, or expressions that return filenames).  Wildcard characters (`*`, which matches everything and `?`, which matches any single character) are acceptable. Nested lists are flattened.\n",
    "\n",
    "Examples of input specification are as follows:\n",
    "\n",
    "```\n",
    "input: []\n",
    "\n",
    "input: 'file1.fasta', 'file2.fasta'\n",
    "\n",
    "input: 'file*.fasta', 'filename with space.fasta'\n",
    "\n",
    "input:\n",
    "    'file*.txt',\n",
    "    'directory/file2.txt'\n",
    "\n",
    "input: aligned_reads\n",
    "\n",
    "input: aligned_reads, reference_genome\n",
    "\n",
    "input: aligned_reads[2:]\n",
    "\n",
    "input: 'data/*.fastq'\n",
    "\n",
    "input: '*/GXT*.fastq'\n",
    "\n",
    "input: func(parameter)\n",
    "```\n",
    "\n",
    "It is worth noting that\n",
    "\n",
    "* The first examples shows that the step does not need any input file (so it does not depend on any other step).\n",
    "* It does not matter if `aligned_reads` and `reference_genome` are strings or lists of strings because SoS will flatten nested lists to a single list of filenames.\n",
    "* The `input` option tries to expand filenames with wildcard characters (`*` and `?`). This can be very useful for workflows that, for example, regularly scan a directory and process unprocessed files. However, because the value of this step depends on availability of files, the output of `sos show script` and the execution path will be unpredictable, and even wrong if there is no available file during the execution of `sos show script`.\n",
    "\n",
    "The input files will be evaluated and form a list of input files. They are by default sent to the step process all at once as varible `_input`, but can also be sent in groups, each time\n",
    "with different `_input`. Here `_input` is a temporary variable that is available only within the step."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `filetype`\n",
    "\n",
    "SoS allows the specification of input options, which are appended to input file list as comma separated `name=value` pairs.\n",
    "\n",
    "Option `filetype` accepts one or more filetypes or a lambda function. For example,\n",
    "\n",
    "```\n",
    "[step]\n",
    "input:\n",
    "\tinput_files, filetype='*.fastq'\n",
    "```\n",
    "\n",
    "passes only files with extension `*.fastq`.\n",
    "\n",
    "```\n",
    "[step]\n",
    "input:\n",
    "\tinput_files,\n",
    "\tfiletype=['*.fastq', '*.fastq.gz']\n",
    "\n",
    "```\n",
    "\n",
    "passes only files with extension `.fastq` or `.fastq.gz`. Under the hood SoS treats the pattern as Unix shell-stype wildcard pattern (with `*`, `?`, `[seq]` and `[!seq]`, see [doc](https://docs.python.org/2/library/fnmatch.html#module-fnmatch) for details) so\n",
    "\n",
    "* **`filetype='.txt'` does not match `file.txt`**\n",
    "* `filetype='*.fastq*'` matches `a.fastq`, `a.fastq.gz` and `a.fastq.zip`\n",
    "* `filetype='[!_]*.txt'` matches `file1.txt` but not `_file1.txt`\n",
    "\n",
    "If you need more refined control over the selection of files, you can use lambda functions (a bit python knowledge is required). For example,\n",
    "\n",
    "```\n",
    "[step]\n",
    "input:\n",
    "\tinput_files,\n",
    "\tfiletype=lambda x: open(x).readline().startswith('##fileformat=VCF4.1')\n",
    "```\n",
    "\n",
    "passes only files with the first line starting with string `##fileformat=VCF4.1``."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `group_by`\n",
    "\n",
    "SoS by default passes all input files to step process as a single list. Option `group_by` pass input files in groups, each time with a subset of input files named `_input`. SoS allows you to group input by `single` (individual file), `pairs` (match first half of files with the second half),  `combinations` (all unordered combinations of 2-sets), `pairwise` (all adjacent 2-sets), or chunks of size `N` (integer `group_by=3` or string `group_by='4'`). For example, with the following sos script\n",
    "\n",
    "```\n",
    "# how to group input files\n",
    "parameter: group = 'all'\n",
    "\n",
    "[0]\n",
    "print('group_by=${group}')\n",
    "input: 'file1', 'file2', 'file3', 'file4', group_by=group\n",
    "print('${_index}: ${_input}')\n",
    "```\n",
    "\n",
    "The output of commands are\n",
    "\n",
    "```\n",
    "$ sos run test.sos  -i -v0\n",
    "group_by=all\n",
    "0: file1 file2 file3 file4\n",
    "\n",
    "$ sos run test.sos --group 'single' -i -v0\n",
    "group_by=single\n",
    "0: file1\n",
    "1: file2\n",
    "2: file3\n",
    "3: file4\n",
    "\n",
    "$ sos run test.sos --group 'pairwise' -i -v0\n",
    "group_by=pairwise\n",
    "0: file1 file2\n",
    "1: file2 file3\n",
    "2: file3 file4\n",
    "\n",
    "$ sos run test.sos --group 'pairs' -i -v0\n",
    "group_by=pairs\n",
    "0: file1 file3\n",
    "1: file2 file4\n",
    "\n",
    "$ sos run test.sos --group 'combinations' -i -v0\n",
    "group_by=combinations\n",
    "0: file1 file2\n",
    "1: file1 file3\n",
    "2: file1 file4\n",
    "3: file2 file3\n",
    "4: file2 file4\n",
    "5: file3 file4\n",
    "\n",
    "$ sos run test.sos --group 1 -i -v0\n",
    "group_by=1\n",
    "0: file1\n",
    "1: file2\n",
    "2: file3\n",
    "3: file4\n",
    "\n",
    "$ sos run test.sos --group 2 -i -v0\n",
    "group_by=2\n",
    "0: file1 file2\n",
    "1: file3 file4\n",
    "```\n",
    "Here we are running the script in inspect mode to avoid \"file not found\" errors for input files.\n",
    "\n",
    "Obviously, the output of the `pairs` cases depends on the order of files. If you need to pair files in any particular order, you can control it in input. For example\n",
    "\n",
    "```\n",
    "[step]\n",
    "input:\n",
    "\tsorted([x for x in fastq_files if '_R1_' in x]),\n",
    "\tsorted([x for x in fastq_files if '_R2_' in x]),\n",
    "\tgroup_by='pairs'\n",
    "\n",
    "run('echo ${input}')\n",
    "```\n",
    "\n",
    "will take all input files and sort them by `_R1_` and `_R2_` and by filename, and pair them.\n",
    "\n",
    "### Option `for_each`\n",
    "\n",
    "Option `for_each` allows you to repeat step process for each value of a variable. For example,\n",
    "\n",
    "```\n",
    "[0]\n",
    "method = ['m1', 'm2']\n",
    "input: 'file1', 'file2', for_each='method'\n",
    "print('${_index}: ${_input} ${_method}')\n",
    "```\n",
    "\n",
    "will repeat the step with each item of variable `method`\n",
    "\n",
    "```\n",
    "$ sos run test.sos -v0\n",
    "0: file1 file2 m1\n",
    "1: file1 file2 m2\n",
    "```\n",
    "\n",
    "Note that the SoS automatically creates a loop variable `_method` for variable `method`. If the variable is an attribute of an object (e.g. `aligned.output`), the iterator variable will be named without the attribute part.\n",
    "\n",
    "Nested loops are also allowed. For example,\n",
    "\n",
    "```python\n",
    "[0]\n",
    "method = ['m1', 'm2']\n",
    "pars = [1, 2]\n",
    "input: 'file1', 'file2', for_each=['method', 'pars']\n",
    "print('${_index}: _input=${_input} _method=${_method}, _pars=${_pars}')\n",
    "```\n",
    "\n",
    "would produce\n",
    "\n",
    "```\n",
    "$ sos run test.sos -v0\n",
    "0: _input=file1 file2 _method=m1, _pars=1\n",
    "1: _input=file1 file2 _method=m2, _pars=1\n",
    "2: _input=file1 file2 _method=m1, _pars=2\n",
    "3: _input=file1 file2 _method=m2, _pars=2\n",
    "```\n",
    "\n",
    "If you would like to loop the process with several parameters, you can put them into the same level by 'var1,var2'. For example,\n",
    "\n",
    "\n",
    "```python\n",
    "[0]\n",
    "method = ['m1', 'm2']\n",
    "pars = [1, 2]\n",
    "input: 'file1', 'file2', for_each='method,pars'\n",
    "print('${_index}: _input=${_input} _method=${_method}, _pars=${_pars}')\n",
    "```\n",
    "\n",
    "would produce\n",
    "\n",
    "```\n",
    "$ sos run test.sos -v0\n",
    "0: _input=file1 file2 _method=m1, _pars=1\n",
    "1: _input=file1 file2 _method=m2, _pars=2\n",
    "```\n",
    "\n",
    "The variable passed to option `for_each` can a sequence, or a Pandas dataframe. In this case, each `_loop` variable presents a line in the dataframe and you can access single values using format `_loop[\"header\"]`. For example\n",
    "\n",
    "```\n",
    "[0]\n",
    "import pandas as pd\n",
    "data = pd.DataFrame([(1, 2, 'Hello'), (2, 4, 'World')], columns=['A', 'B', 'C'])\n",
    "input: for_each='data'\n",
    "output: '${_data[\"A\"]}_${_data[\"B\"]}_${_data[\"C\"]}.txt'\n",
    "sh:\n",
    "    touch ${_output}\n",
    "```\n",
    "\n",
    "would produce results with names `1_2_Hello.txt` and `2_4_World.txt`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `paired_with`\n",
    "\n",
    "Input files might might come with additional information such as sample type, and sample name, and you can pair these information to input files using the `paired_with` option. For example, `bam_files` in the following example have matched `mutated` and `sample_name`, you can atteched these information to groups of input files `_input` with looped paired with variables.\n",
    "\n",
    "```\n",
    "[0]\n",
    "bam_files = ['case/A1.bam', 'case/A2.bam', 'ctrl/A1.bam', 'ctrl/A2.bam']\n",
    "mutated = ['case', 'case', 'ctrl', 'ctrl']\n",
    "sample_name = ['A1', 'A2', 'A1', 'A2']\n",
    "\n",
    "input: bam_files, paired_with=['mutated', 'sample_name'], group_by='pairs'\n",
    "\n",
    "print('${_index}: _input=${_input} _mutated=${_mutated}, _sample_name=${_sample_name}')\n",
    "```\n",
    "\n",
    "Output:\n",
    "\n",
    "```\n",
    "$ sos run test.sos -v0\n",
    "0: _input=case/A1.bam ctrl/A1.bam _mutated=case ctrl, _sample_name=A1 A1\n",
    "1: _input=case/A2.bam ctrl/A2.bam _mutated=case ctrl, _sample_name=A2 A2\n",
    "```\n",
    "\n",
    "Similar to option `for_each`, if the variable is an attribute of an object (e.g. `aligned.output`), the iterator variable will be named without the attribute part (e.g. `_aligned` for `paired_with='aligned.output'`)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option `pattern`\n",
    "\n",
    "This option does the reverse of function `expand_pattern`. It uses named wildcards to match pattern to all input files, and creates step variables for these wildcard objects. For example,\n",
    "\n",
    "```python\n",
    "[step]\n",
    "input:  'a-20.txt', 'b-10.txt', pattern = '{name}-{par}.txt'\n",
    "output: expand_pattern('{name}-processed-{par}.txt')\n",
    "run('echo ${output}; touch ${output}')\n",
    "```\n",
    "\n",
    "will take all input files and extract `name` and `par` from each file name as variables `name` and `par`. It is then used to create output file names adding the word `processed` in between these wildcard objects. The outcome of the SoS script above is creation of files `a-processed-10.txt` and `b-processed-20.txt`.\n",
    "\n",
    "When wildcard objects are accessed as step variables, both variable names with and without `_` prefix is available, e.g. in this example, both `_name` and `name`, `_par` and `par` are avaiable and are the same. The two conventions will only differ when [`group_by`](Documentation#option-group_by) or ['for_each'](Documentation#option-for_each) is also used.\n",
    "\n",
    "```python\n",
    "[step]\n",
    "input:  'a-20.txt', 'b-10.txt', extract = '{name}-{par}.txt', group_by='single'\n",
    "output: expand = '{_name}-processed-{_par}.txt'\n",
    "run('echo ${_output}; touch ${_output}')\n",
    "```\n",
    "\n",
    "Please note that the sigil `{}` is exclusively used for named wildcards and you should not use `${}` because it will be interpreted as strings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `dynamic` input files\n",
    "\n",
    "In order to determine the best execution strategy, SoS evaluates all expressions for all steps before the execution of a workflow to figure\n",
    "out input and output of steps. This works most of the time but sometimes the input of a step can only be determined at runtime. For example,\n",
    "if you would like your workflow to automatically scan an input directory and process all fasta files under it, or if a previous step produces\n",
    "files that cannot be determined beforehand, you can specify input files as follows,\n",
    "\n",
    "```python\n",
    "input: 'input/*.fasta'\n",
    "```\n",
    "\n",
    "The problem is that no file or a wrong set files might exist during the planing stage so SoS might skip this step or start the step\n",
    "with a wrong set of files. To address this problem, you can declare the input files as **dynamic** by passing a `dynamic` object\n",
    "\n",
    "```python\n",
    "input: dynamic('input/*.fasta')\n",
    "```\n",
    "\n",
    "This tells SoS that the input of this step can only be determined at runtime and will execute the step only after all its previous\n",
    "steps have been completed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary\n",
    "\n",
    "Options of step `input` are evaluated in the following orders:\n",
    "\n",
    "1. A list of input files, if specified, would replace `input`, which is by default output from the previous step.\n",
    "2. Option `filetype` filters input files. **The output becomes `input`**.\n",
    "3. Option `group_by` groups the files into several groups, named `_input`\n",
    "4. Option `for_each` repeat `_input` for each loop var, named `_loopvar` if `for_each='loopvar'`.\n",
    "5. Option `paired_with` pairs one or more variables with `input`, variable `paired` is paired with `input`\n",
    "    and variable `_paired` is paired with `_input` in each loop if `paired_with='paired'`\n",
    "6. Option `pattern` extract variables from filenames in `input`. Variable `extracted` is paired with `input`\n",
    "    and variable `_extracted` is paired with `_input` in each loop if `extract='{extracted}_other_part'`.\n",
    "7. Option `skip` optionally skip all or part of the input groups.\n",
    "\n",
    "The differences between looped and non-loop steps are sumarized in the following figure\n",
    "\n",
    "![step_loop](../media/step_loop.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output files\n",
    "\n",
    "Output files of a step can be specified by step `output`. Whereas step `input` override or\n",
    "change SoS provided variable `input` to produce one or more variable `_input`,\n",
    "the step `output` specify `output` which should be `[]` if no output is generated\n",
    "(or of interest). Similar to `input`, step output accepts strings, variables, expressions, and allows wildcard characters. For example, the following are acceptable output files\n",
    "\n",
    "```python\n",
    "output:  []\n",
    "\n",
    "output:  'accepted_hits.bam'\n",
    "\n",
    "output:  aligned_reads, bam_stats\n",
    "\n",
    "output:  'aligned/*.bam'\n",
    "\n",
    "output:  expand_pattern('aligned_{samples}.bam')\n",
    "```\n",
    "\n",
    "In the last example, function `expand_pattern` is used to contruct list of files from items of a sequence `samples`. In case of `input` loop, step `output` actually determines variable `_output` for each input loop. For example, the following step accepts one or more bam files and index them using command `samtools index`. The input files are passed one\n",
    "by one so generate multiple `output` files (`'${_input}.bai'`). \n",
    "\n",
    "```python\n",
    "[10]\n",
    "input:\n",
    "\tbamfiles, group_by='single'\n",
    "\n",
    "output:\n",
    "\t'${_input}.bai'\n",
    "\n",
    "run('''samtools index ${_input} ''')\n",
    "```\n",
    "\n",
    "The use of variable `output` in this scenario is discouraged because `output`, as the collection of all `_output` increases with each input group. The accumulation only happens when `_output` and `output` are different so `output` will stay the same if all output files are specified as in the following example\n",
    "\n",
    "```python\n",
    "output: [x + '.bai' for x in bamfiles]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `dynamic` output files\n",
    "\n",
    "Similar to the cases with [dynamic input files](#dynamically-determined-input-files-function-dynamic), the\n",
    "output of some steps could also not be determined beforehand. For example, with the following script that generates `html` files that cannot be determined during dry run, \n",
    "\n",
    "```python\n",
    "[0]\n",
    "run:\n",
    "    rm -f *.html\n",
    "\n",
    "[10]\n",
    "input: []\n",
    "output: '*.html'\n",
    "\n",
    "import random\n",
    "for i in range(5):\n",
    "    run('touch result_${random.randint(1, 20)}.html')\n",
    "```\n",
    "\n",
    "SoS will determine that you do not have any output file (no `*.html` file) and produce the following output\n",
    "\n",
    "```bash\n",
    "$ sos run test.sos\n",
    "INFO: Execute default_0:\n",
    "INFO: input:   []\n",
    "INFO: output:  unspecified\n",
    "INFO: Execute default_10:\n",
    "INFO: input:   []\n",
    "WARNING: *.html does not expand to any valid file.\n",
    "INFO: output:  []\n",
    "```\n",
    "\n",
    "In this case, you will need to define the output as `dynamic` using\n",
    "\n",
    "```\n",
    "[0]\n",
    "run:\n",
    "    rm -f *.html\n",
    "\n",
    "[10]\n",
    "input: []\n",
    "output: dynamic('*.html')\n",
    "\n",
    "import random\n",
    "for i in range(5):\n",
    "    run('touch result_${random.randint(1, 20)}.html')\n",
    "```\n",
    "\n",
    "so that SoS knows that the output can only be determined after the completion of the step. Because of this, variable `output` is unavailable to the step process."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step depends"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This item specifies files that are required for the step. Although not required, it is a good practice to list resource files and other dependency files for a particular step. For example\n",
    "\n",
    "```python\n",
    "[10]\n",
    "input:\n",
    "\tfasta_files\n",
    "\n",
    "depends:\n",
    "\treference_seq\n",
    "\n",
    "```\n",
    "\n",
    "Similar to `output` options, dependent files can also be defined after `input` options and consist of\n",
    "dependent files determined from loop variables. \n",
    "\n",
    "The following figure summarizes the effect of `input`\n",
    "and `output` options and input options `group_by` and `for_each` on the flow\n",
    "of input and output files and related variables.\n",
    "\n",
    "![Step options](../media/step_options.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step process"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A step process is the Python statements that perform certain tasks and produce step output from step input. A step process can contain arbitrary Python statements. For example,\n",
    "\n",
    "```\n",
    "[10]\n",
    "output: 'a.txt'\n",
    "with open(${_output!r}, 'w') as dest:\n",
    "   dest.write('some text')\n",
    "```\n",
    "\n",
    "and\n",
    "\n",
    "```\n",
    "[10]\n",
    "output: 'a.txt'\n",
    "run('echo \"some text\" > ${_output!q}')\n",
    "```\n",
    "\n",
    "use inline (interpreted and executed by SoS) python code or shell script to generate `a.txt`. \n",
    "\n",
    "Step processes are executed within SoS and are executed sequentially. However, part or all of the step process can be executed externally and potentially in parallel as step `task`."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos.kernel.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "toc": {
   "nav_menu": {
    "height": "503px",
    "width": "254px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
